src/main/scala/ftanml/util/Implicits.scala

```
@@ -0,0 +1,40 @@
+package ftanml.util
+
+import ftanml.objects._
+
+object Implicits {
+	
+	class NotAnInstanceException(message: String) extends IllegalArgumentException(message)
+	
+	implicit def FtanValue2FtanNumber(value: FtanValue) = {
+		if(value.isInstanceOf[FtanNumber]){
+			value.asInstanceOf[FtanNumber]
+		}else{
+			throw new NotAnInstanceException("Value " + value + " is not an instance of FtanNumber")
+		}
+	}
+	
+	implicit def FtanValue2FtanString(value: FtanValue) = {
+		if(value.isInstanceOf[FtanString]){
+			value.asInstanceOf[FtanString]
+		}else{
+			throw new NotAnInstanceException("Value " + value + " is not an instance of FtanString")
+		}
+	}
+	
+	implicit def FtanValue2FtanArray(value: FtanValue) = {
+		if(value.isInstanceOf[FtanArray]){
+			value.asInstanceOf[FtanArray]
+		}else{
+			throw new NotAnInstanceException("Value " + value + " is not an instance of FtanArray")
+		}
+	}
+	
+	implicit def FtanValue2FtanElement(value: FtanValue) = {
+		if(value.isInstanceOf[FtanElement]){
+			value.asInstanceOf[FtanElement]
+		}else{
+			throw new NotAnInstanceException("Value " + value + " is not an instance of FtanElement")
+		}
+	}
+}
```

I don't think they should do too much harm. They just let you omit the asInstanceOf in places where the subtype needed can be inferred. They don't do anything different of asInstanceOf then, only trying to give a nicer error message in case that fails. Of course we could omit that special error handling or do some more sophisticated error handling here without having to explicitly do it where we need such casts.

A function would not solve the problem. asInstanceOf already does that (though throwing an error if it's not possible instead of returning None - which is what we probably want in that case, we would have to check for None in lots of places then), this just saves us some writing when inferring the type is possible. Feel free to pull the other commits only if you don't like this one, though, they should not really depend on it.